src/Mvc/Mvc.DataAnnotations/src/DataAnnotationsMetadataProvider.cs

```diff
@@ -366,10 +366,27 @@ public void CreateValidationMetadata(ValidationMetadataProviderContext context)
                 }
                 else if (context.Key.MetadataKind == ModelMetadataKind.Property)
                 {
-                    addInferredRequiredAttribute = IsNullableReferenceType(
-                        context.Key.ContainerType,
-                        member: null,
-                        context.PropertyAttributes);
+                    var property = context.Key.PropertyInfo;
+                    if (property is null)
+                    {
+                        // PropertyInfo was unavailable on ModelIdentity prior to 3.1. 
+                        // Making a cogent argument about the nullability of the property requires inspecting the declared type,
+                        // since looking at the runtime type may result in false positives: https://github.com/aspnet/AspNetCore/issues/14812
+                        // The only way we could arrive here is if the ModelMetadata was constructed using the non-default provider.
+                        // We'll cursorily examine the attributes on the property, but not the ContainerType to make a decision about it's nullability.
+
+                        if (HasNullableAttribute(context.PropertyAttributes, out var propertyHasNullableAttribute))
+                        {
+                            addInferredRequiredAttribute = propertyHasNullableAttribute;
```

Comment:
Hmm, i'm confused, so if it has the nullable attribute then we DO add the required attribute? Isn't that backwards? I'm sure it's just me misunderstanding the code since your mass amount of tests show differently.

Reply:
This might help: https://github.com/dotnet/roslyn/blob/master/docs/features/nullable-metadata.md#type-parameters, but essentially `non-null` is represented as `[Nullable(1)]`. `HasNullableAttribute` returns `true` if the attribute is present, the `propertyHasNullableAttribute` says if the value is `1`.